### DISCUSSION TITLE
relax CS8801 for const values in same file

### DISCUSSION THREAD

**Discussion by ZacharyPatten** [dotnet/csharplang#6144] [Language Ideas]
I think `CS8801` is fine for non-`const` values, but it seems a bit too much for `const` values in the same file. Perhaps `const` values should not trigger `CS8801` or perhaps there should be a way around the error. The following code does not currently compile:
```cs
using System;

const string a = "foo";
const string b = "bar";
const string c = "foobar";

enum MyEnum
{
    [My(a)] A, // <- CS8801
    [My(b)] B, // <- CS8801
    [My(c)] C, // <- CS8801
}

class MyAttribute : Attribute
{
    public string FooBar { get; set; }
    public MyAttribute(string fooBar) => FooBar = fooBar;
}
``` 

In order to make the above code compile currently, we would need to wrap the `const` values in a type. Perhaps something like:
```cs
using System;
using static MyConsts;

Console.WriteLine("top level statement");

static class MyConsts
{
    public const string a = "foo";
    public const string b = "bar";
    public const string c = "foobar";
}

enum MyEnum
{
    [My(a)] A,
    [My(b)] B,
    [My(c)] C,
}

class MyAttribute : Attribute
{
    public string FooBar { get; set; }
    public MyAttribute(string fooBar) => FooBar = fooBar;
}
```
The need to do this for `const` values seems a bit excessive to me.

I admit it may confuse people if `const` variables didn't trigger `CS8801` but non-`const` variables did, but I think it is worth considering.

Another example that currently doesn't compile:
```cs
using System;

const string a = "foo";

class MyClass
{
    public static void MyMethod() => Console.WriteLine(a); // <- CS8801
}
```

**Comment by HaloFour** [comment]
Constants in .NET are still fields which I believe must be associated with a type:

```cs
const string a = "foo";
```

```
.field private static literal string a = "foo"
```

**Comment by jnm2** [comment]
Can you say more about why placing constants in a static class is a bit too much? Names like a, b, and c don't demonstrate this.

**Reply by ZacharyPatten** [reply]
> why placing constants in a static class is a bit too much

Just my opinion. The whole point of top level statements is to reduce the boiler plate. Having to put it inside a type just works against the idea of top level statements. `const` values "may not be modified" (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/const), so why not just allow us to reference them in other scopes/types in the same file (essentially treat it as if it was a field rather than a local)?

**Reply by jnm2** [reply]
It *is* consistent with top level statements in that, anything declared in top level statements is not accessible inside any declared types in the project. Top-level ordinary variables and local functions can't be called outside of top-level statements either, and it would be a big deal to keep these three things consistent (see the threads on free functions).

**Reply by HaloFour** [reply]
You can already declare `const` locals in top-level statements:

```cs
const string FOO = "Foo";
System.Console.WriteLine(FOO);
```

But like anything else in top-level statements those belong to the synthesized `Main` method and their scope doesn't extend to types declared within that same file.

IMO this belongs in the same discussion as top-level members:

https://github.com/dotnet/csharplang/issues/2765
